src/server/index.js

```diff
@@ -1,188 +1,244 @@
-      conn = await newMuxConn(muxed) // get a muxed connection
+  addToNetwork (connection, wrappedStream, id) {
+    log('adding %s to network', id.toB58String())
 
-      conn.muxed = muxed
-      await this.switch.negotiateProtocol(conn)
-    } catch (err) {
-      log(err)
+    const client = this.network[id.toB58String()] = {
+      id,
+      connection,
+      wrappedStream
     }
+
+    checkAckLoop(wrappedStream, () => this.removeFromNetwork(client))
+
+    this.update()
+    this.broadcastDiscovery()
   }
 
-  async handlerV010 (conn, muxed) {
-    const rpc = LP(conn) // turn into length-prefixed rpc interface
+  /**
+   * Update discovery data
+   */
+  update () {
+    log('updating cached data')
+    this.networkArray = Object.values(this.network)
+    this._cachedDiscovery = this.networkArray.length ? { ids: this.networkArray.map(client => client.id._id) } : { ids: [] }
+  }
 
-    try {
-      log('performing challenge')
+  /**
+   * Broadcast discovery data
+   */
+  broadcastDiscovery () {
+    log('broadcasting discovery to %o client(s)', this.networkArray.length)
+    this.networkArray.forEach(client => {
+      client.wrappedStream.writePB(this._cachedDiscovery, Discovery)
+    })
+  }
 
-      const { random128: random, peerID } = await rpc.readProto(JoinInit)
-      const id = await prom(cb => ID.createFromJSON(peerID, cb))
+  async _register (random, id, wrappedStream, stream) {
+    log('performing challenge')
 
-      if (!Buffer.isBuffer(random) || random.length !== 128) {
-        rpc.writeProto(JoinVerify, { error: Error.E_RAND_LENGTH })
-        return muxed.end()
-      }
+    if (!Buffer.isBuffer(random) || random.length !== 128) {
+      wrappedStream.writePB({ error: Error.E_RAND_LENGTH }, JoinVerify)
 
-      log('got id, challenge for %s', id.toB58String())
+      // close the stream, no need to wait
+      stream.sink([])
+      return
+    }
 
-      const saltSecret = crypto.randomBytes(128)
-      const saltEncrypted = await prom(cb => id.pubKey.encrypt(saltSecret, cb))
+    log('got id, challenge for %s', id.toB58String())
 
-      rpc.writeProto(JoinChallenge, { saltEncrypted })
+    const saltSecret = crypto.randomBytes(128)
+    const saltEncrypted = await id.pubKey.encrypt(saltSecret)
 
-      const solution = sha5(random, saltSecret)
-      const { solution: solutionClient } = await rpc.readProto(JoinChallengeSolution)
+    await wrappedStream.writePB({ saltEncrypted }, JoinChallenge)
 
-      if (solution.toString('hex') !== solutionClient.toString('hex')) {
-        rpc.writeProto(JoinVerify, { error: Error.E_INCORRECT_SOLUTION })
-        return muxed.end()
-      }
+    const solution = sha5(random, saltSecret)
+    const { solution: solutionClient } = await wrappedStream.readPB(JoinChallengeSolution)
 
-      rpc.writeProto(JoinVerify, {})
+    if (solution.toString('hex') !== solutionClient.toString('hex')) {
```

Comment:
If anyone knows a better method to directly compare buffers it would be great, this is more of a hack from me

Reply:
Any reason for not using https://nodejs.org/api/buffer.html#buffer_buf_compare_target_targetstart_targetend_sourcestart_sourceend ?

I tried to keep the logic the same to not make the refactor noisy

Reply:
Me not knowing this exists :P

Reply:
ahah, I will change it then :)